man/man1/swin.1
.\" @(#)swin.1 1.1 92/07/30 SMI;
.TH SWIN 1 "21 December 1987"
.SH NAME
swin \- set or get SunView user input options
.SH SYNOPSIS
.B swin
[
.B \-cghm
]
[
.B \-r
.I event
.I value
.I shift_state
]
[
.B \-s
.I event
.I value
.I shift_state
]
[
.B \-t
.I seconds
]
.SH AVAILABILITY
.LP
This command is available with the
.TX SVBG 
software installation option.  Refer to
.TX INSTALL
for information on how to install optional software.
.SH DESCRIPTION
.IX swin "" "\fLswin\fR \(em set window input behavior"
.LP
The
.B swin
(set window; analogous to
.BR stty (1V))
command lets you change some of the input behavior of your SunView
environment.
By default, your keyboard input follows your pointer.
This means that in order to type to a window you position the 
pointer over the window.  This is called
.I keyboard-follows-mouse
mode.
.LP
You can specify that the keyboard
input continues to go to the same window,
regardless of the pointer position,
until you take some specific action, like clicking the mouse.
When this is done, you can roam around the screen with the 
pointer and not change the window to which keyboard input is directed.
Running SunView like this is said to be operating in
.I click-to-type
mode.
.LP
When running in click-to-type mode, one user action
.I sets
the type-in point in the window that you want to receive keyboard input.
The default user action to do this is the clicking of the
.SM LEFT
mouse button while positioning the pointer
over the new type-in point.
This user action can be changed.
.LP
Another user action
.I restores
the previous type-in point in the window that
you want to receive keyboard input.
The default user action to do this is the clicking of the
.SM MIDDLE
mouse button while positioning the pointer
over the window.  This user action can be changed.
.SH OPTIONS
.TP
.B \-c
Turn on click-to-type mode using the default user actions: the
.SM LEFT
mouse button sets the type-in point and the
.SM MIDDLE
mouse button restores
the type-in point.  You can use the
.BR defaultsedit (1)
program to set
click-to-type on permanently; see the
.B Click_to_Type
option of
.BR sunview (1).
.TP
.B \-g
Get the state of the user input options controlled by
.BR swin .
If no arguments are supplied to
.B swin
then
.B \-g
is implied.
.TP
.B \-h
Print out a help message that briefly describes the options to
.BR swin .
.TP
.B \-m
Run in keyboard-follows-mouse mode.
.TP
.BI \-s " event value shift_state"
Set the user action that sets the type-in point and
sets the keyboard input window.  The
.I event
identifies the particular user action and is one of:
.RS
.TP
.SM LOC_WINENTER
pointer entering a window
.TP
.SM MS_LEFT
.SM LEFT
mouse button
.TP
.SM MS_MIDDLE
.SM MIDDLE
mouse button
.TP
.SM MS_RIGHT
.SM RIGHT
mouse button
.br
.ne 5
.TP
.I decimal_number
place the decimal number of a firm event here;
see list of events in
.B <sundev/vuid_event.h>
(avoid function keys, normally unused control-\s-1ASCII\s0
characters are
.SM OK\s0,
normally unused
.SM SHIFT
keys are
.SM OK\s0).
.LP
.I value
identifies the transition of the
.I event
and is one of:
.TP
.SM ENTER
the pointer entering a window (use with
.SM LOC_WINENTER\s0)
.TP
.SM DOWN
the button associated with
.I event
went down
.TP
.SM UP
the button associated with
.I event
went up (avoid this)
.br
.ne 6
.LP
The
.I shift_state
identifies the state of the 
.SM SHIFT
keys at the time of the
.IR event / value
pair in order for that pair to be used to
control the keyboard input window.  The
.I shift_state
is one of:
.TP
.SM SHIFT_DONT_CARE
Ignore the state of the 
.SM SHIFT
keys
.TP
.SM SHIFT_ALL_UP
All the 
.SM SHIFT
keys must be up
.TP
.SM SHIFT_LEFT
The left
.SM SHIFT
key must be down (not the key labeled
.SM LEFT\s0)
.TP
\s-1SHIFT_RIGHT\s0
the right 
.SM SHIFT
key must be down (not the key labeled
\s-1RIGHT\s0)
.TP
\s-1SHIFT_LEFTCTRL\s0
the left 
.SM CTRL
key must be down
.TP
\s-1SHIFT_RIGHTCTRL\s0
the right 
.SM CTRL
key must be down
.RE
.TP
.BI \-r " event value shift_state"
Set the user action that restores the type-in point and
sets the keyboard input window.
This user action is swallowed so that the application that owns the window
does not see it.  However, if the window already has keyboard input or
if the window refuses keyboard input then
this user action is passed on through to the application.
The parameters to this command are like those for
.BR \-s .
The following example shows modifying the default click-to-type user
actions so that a
.SM SHIFT
left is required for the restore user event:
.RS
.IP
.B
example% swin \-c \-r \s-1MS_MIDDLE DOWN SHIFT_LEFT\s0
.RE
.TP
.BI \-t " seconds"
SunView synchronizes input so that it does not hand out the next user
action until the application fielding the current user action finishes
its processing.  This allows type-ahead and mouse-ahead.
If an application does not finish processing within a given
length of time (process virtual time; not wall clock time),
the next user action is handed out anyway.
This avoids any one application from hanging the workstation.
The
.B \-t
command sets this time limit.
A
.I seconds
value of 0 tells SunView to run unsynchronized;
beware of race conditions in this mode.  The default seconds value
is 2 and the
.B \-c
command makes it 10 seconds.
.SH SEE ALSO
.BR defaultsedit (1),
.BR stty (1V),
.BR sunview (1)
.LP
.TX SVBG
.br
.ne 5
.SH DIAGNOSTICS
.TP
.B "swin not passed parent window in environment"
.B swin
does not work unless SunView is started already.
.br
.ne 5
.SH BUGS
.LP
.B swin
gets you no help in preventing you from specifying
.B \-r
or
.B \-s
parameters that are not sensible.


usr.bin/sunwindows/utilities/swin.c
#ifndef lint
#ifdef sccs
static	char sccsid[] = "@(#)swin.c 1.1 92/07/30";
#endif
#endif

/*
 * Sun Microsystems, Inc.
 */

/*
 * 	Overview:	Set/get runtime SunView controls.
 *			This is supposed to be the window equivalent of stty.
 */

#include <stdio.h>
#include <sunwindow/window_hs.h>
#include <sys/file.h>

#ifdef STANDALONE
#define EXIT(n)		exit(n)
#else
#define EXIT(n)		return(n)
#endif

#define	DONT_CARE_SHIFT		-1
#define	SHIFT_MASK(bit) (1 << (bit))

#ifdef STANDALONE
main(argc, argv)
#else
swin_main(argc, argv)
#endif
	int argc;
	char **argv;
{
	int	parent_fd;
	char	parentname[WIN_NAMESIZE];
	char	*prog_name = *argv;
	Firm_event fe_focus;
	int fe_focus_shift = DONT_CARE_SHIFT;
	Firm_event fe_restore_focus;
	int fe_restore_focus_shift;
	struct timeval tv;

	/* Determine parent */
	if (we_getparentwindow(parentname)) {
		(void)fprintf(stderr,
		    "%s not passed parent window in environment\n", prog_name);
		EXIT(1);
	}
	/* Open parent window */
	if ((parent_fd = open(parentname, O_RDONLY, 0)) < 0) {
		(void)fprintf(stderr, "%s (parent) would not open\n", parentname);
		EXIT(1);
	}
	argv++;
	argc--;
	if (argc == 0)
		show_settings(parent_fd);
	/* Get application related args */
	while (argc > 0 && **argv == '-') {
		switch (argv[0][1]) {

		case 'c': 	/* Click-to-type */
			fe_focus.id = MS_LEFT;
			fe_focus.value = 1;
			fe_focus_shift = DONT_CARE_SHIFT;
			if (win_set_focus_event(parent_fd, &fe_focus,
			    fe_focus_shift)) {
				perror("win_set_focus_event");
				EXIT(1);
			}
			fe_restore_focus.id = MS_MIDDLE;
			fe_restore_focus.value = 1;
			fe_restore_focus_shift = DONT_CARE_SHIFT;
			if (win_set_swallow_event(parent_fd,
			    &fe_restore_focus, fe_restore_focus_shift)) {
				perror("win_set_swallow_event");
				EXIT(1);
			}
			tv.tv_usec = 0;
			tv.tv_sec = 10;
			if (win_set_event_timeout(parent_fd, &tv)) {
				perror("win_set_event_timeout");
				EXIT(1);
			}
			break;

		case 'g':	/* Get state */
			show_settings(parent_fd);
			break;

		case 'm':	/* Keyboard follows 'm'ouse */
			fe_focus.id = LOC_WINENTER;
			fe_focus.value = 1;
			fe_focus_shift = DONT_CARE_SHIFT;
			if (win_set_focus_event(parent_fd, &fe_focus,
			    fe_focus_shift)) {
				perror("win_set_focus_event");
				EXIT(1);
			}
			fe_restore_focus.id = LOC_WINENTER;
			fe_restore_focus.value = 1;
			fe_restore_focus_shift = DONT_CARE_SHIFT;
			if (win_set_swallow_event(parent_fd,
			    &fe_restore_focus, fe_restore_focus_shift)) {
				perror("win_set_swallow_event");
				EXIT(1);
			}
			break;

		case 'r':	/* Set focus restorer */
			get_focus_from_args(&argc, &argv, (short *)&fe_restore_focus.id,
			    &fe_restore_focus.value, &fe_restore_focus_shift);
			if (win_set_swallow_event(parent_fd,
			    &fe_restore_focus, fe_restore_focus_shift)) {
				perror("win_set_swallow_event");
				EXIT(1);
			}
			break;

		case 's':	/* Set focus changer */
			get_focus_from_args(&argc, &argv, (short *)&fe_focus.id,
			    &fe_focus.value, &fe_focus_shift);
			if (win_set_focus_event(parent_fd, &fe_focus,
			    fe_focus_shift)) {
				perror("win_set_focus_event");
				EXIT(1);
			}
			break;

		case 't': {	/* Set event lock timeout */
			if (argc < 2) {
				(void)fprintf(stderr, "-t takes a seconds arg\n");
				EXIT(1);
			}
			argv++;
			argc--;
			tv.tv_usec = 0;
			tv.tv_sec = atoi(*argv);
			if (win_set_event_timeout(parent_fd, &tv)) {
				perror("win_set_event_timeout");
				EXIT(1);
			}
			break;
			}

		case 'h':	/* Help */
		case 'H':	/* Help */
		case '?':	/* Help */
			usage();
			EXIT(0);
		default:
			;
		}
		argv++;
		argc--;
	}
	EXIT(0);
}

static
show_settings(parent_fd)
	int	parent_fd;
{
	Firm_event focus;
	int shift;
	struct timeval tv;

	if (win_get_focus_event(parent_fd, &focus, &shift)) {
		perror("win_get_focus_event");
		exit(1);
	}
	show_focus(focus, shift, "keyboard focus set event");
	if (win_get_swallow_event(parent_fd, &focus, &shift)) {
		perror("win_get_swallow_event");
		exit(1);
	}
	show_focus(focus, shift, "keyboard focus restore event");
	if (win_get_event_timeout(parent_fd, &tv)) {
		perror("win_get_event_timeout");
		exit(1);
	}
	(void)fprintf(stderr, "Event timeout secs: %ld\n", tv.tv_sec);
}

static
usage()
{
	(void)fprintf(stderr, "Options are\t-c (set `c'lick-to-type)\n");
	(void)fprintf(stderr,
	  "\t\t-g (`g'et the state of settable options)\n");
	(void)fprintf(stderr, "\t\t-h (`h'elp)\n");
	(void)fprintf(stderr, "\t\t-m (keyboard follows `m'ouse)\n");
	(void)fprintf(stderr,
	  "\t\t-r event value shift_state (`r'estore kbd focus event)\n");
	(void)fprintf(stderr,
	  "\t\t-s event value shift_state (`s'et kbd focus event)\n");
	(void)fprintf(stderr, "\t\t-t secs (event lock `t'imeout)\n");
	(void)fprintf(stderr,
	  "Event samples:\tLOC_WINENTER, MS_LEFT, MS_MIDDLE\n");
	(void)fprintf(stderr, "Value samples:\tDOWN, ENTER, UP\n");
	(void)fprintf(stderr,
	  "Shift_state samples:\tSHIFT_DONT_CARE, SHIFT_ALL_UP, SHIFT_LEFT\n");
}

static
get_focus_from_args(argc_ptr, argv_ptr, event, value, shift)
	int *argc_ptr;
	char ***argv_ptr;
	register short *event;
	register int *value;
	register int *shift;
{
	char str[200];
	register char *arg;

	if (*argc_ptr < 4) {
		(void)fprintf(stderr, "%s arg count error\n", *argv_ptr);
		usage();
		exit (-1);
	}
	(*argc_ptr)--;
	(*argv_ptr)++;
	arg = **argv_ptr;
	if (strcmp(arg, "LOC_WINENTER") == 0)
		*event = LOC_WINENTER;
	else if (strcmp(arg, "MS_LEFT") == 0)
		*event = MS_LEFT;
	else if (strcmp(arg, "MS_MIDDLE") == 0)
		*event = MS_MIDDLE;
	else if (strcmp(arg, "MS_RIGHT") == 0)
		*event = MS_RIGHT;
	else if (sscanf(arg, "BUT%s", str) == 1)
		*event = atoi(str)+BUT_FIRST;
	else if (sscanf(arg, "KEY_LEFT%s", str) == 1)
		*event = atoi(str)+KEY_LEFTFIRST-1;
	else if (sscanf(arg, "KEY_RIGHT%s", str) == 1)
		*event = atoi(str)+KEY_RIGHTFIRST-1;
	else if (sscanf(arg, "KEY_TOP%s", str) == 1)
		*event = atoi(str)+KEY_TOPFIRST-1;
	else if (strcmp(arg, "KEY_BOTTOMLEFT") == 0)
		*event = KEY_BOTTOMLEFT;
	else if (strcmp(arg, "KEY_BOTTOMRIGHT") == 0)
		*event = KEY_BOTTOMRIGHT;
	else
		*event = atoi(arg);
	(*argc_ptr)--;
	(*argv_ptr)++;
	arg = **argv_ptr;
	if (strcmp(arg, "DOWN") == 0 || strcmp(arg, "Down") == 0 ||
	    strcmp(arg, "down") == 0)
		*value = 1;
	else if (strcmp(arg, "ENTER") == 0 || strcmp(arg, "Enter") == 0 ||
	    strcmp(arg, "enter") == 0)
		*value = 1;
	else if (strcmp(arg, "UP") == 0 || strcmp(arg, "Up") == 0 ||
	    strcmp(arg, "up") == 0)
		*value = 0;
	else
		*value = atoi(arg);
	(*argc_ptr)--;
	(*argv_ptr)++;
	arg = **argv_ptr;
	if (strcmp(arg, "SHIFT_LEFT") == 0)
		*shift = SHIFT_MASK(LEFTSHIFT);
	else if (strcmp(arg, "SHIFT_RIGHT") == 0)
		*shift = SHIFT_MASK(RIGHTSHIFT);
	else if (strcmp(arg, "SHIFT_LEFTCTRL") == 0)
		*shift = SHIFT_MASK(LEFTCTRL);
	else if (strcmp(arg, "SHIFT_RIGHTCTRL") == 0)
		*shift = SHIFT_MASK(RIGHTCTRL);
	else if (strcmp(arg, "SHIFT_DONT_CARE") == 0)
		*shift = DONT_CARE_SHIFT;
	else if (strcmp(arg, "SHIFT_ALL_UP") == 0)
		*shift = 0;
	else
		*shift = atoi(arg);

}

static
show_focus(focus, shift, tag)
	Firm_event	focus;
	int		shift;
	char		*tag;
{
	(void)fprintf(stdout, "State of %s:\n", tag);
	(void)fprintf(stdout, "\tId: ");
	switch (focus.id) {
	case LOC_WINENTER:
		(void)fprintf(stdout, "LOC_WINENTER\n");
		break;

	case MS_LEFT:
		(void)fprintf(stdout, "MS_LEFT\n");
		break;

	case MS_MIDDLE:
		(void)fprintf(stdout, "MS_MIDDLE\n");
		break;

	default: 
		(void)fprintf(stdout, "%ld\n", focus.id);
		break;
	}
	(void)fprintf(stdout, "\tValue: ");
	switch (focus.value) {
	case 0:
		(void)fprintf(stdout, "UP\n");
		break;

	case 1:
		(void)fprintf(stdout, "DOWN or ENTER\n");
		break;

	default:
		(void)fprintf(stdout, "%ld\n", focus.value);
	}
	(void)fprintf(stdout, "\tShift_state: ");
	switch (shift) {
	case DONT_CARE_SHIFT:
		(void)fprintf(stdout, "SHIFT_DONT_CARE\n");
		break;

	case SHIFT_MASK(LEFTSHIFT):
		(void)fprintf(stdout, "SHIFT_LEFT\n");
		break;

	case SHIFT_MASK(LEFTCTRL):
		(void)fprintf(stdout, "SHIFT_LEFTCTRL\n");
		break;

	case SHIFT_MASK(RIGHTSHIFT):
		(void)fprintf(stdout, "SHIFT_RIGHT\n");
		break;

	case SHIFT_MASK(RIGHTCTRL):
		(void)fprintf(stdout, "SHIFT_RIGHTCTRL\n");
		break;

	case 0:
		(void)fprintf(stdout, "SHIFT_ALL_UP\n");
		break;

	default:
		(void)fprintf(stdout, "%lx\n", shift);
	}
}

